e3sm_to_cmip/_logger.py

```diff
@@ -1,68 +1,66 @@
-    """
-    os.makedirs("logs", exist_ok=True)
-    filename = f'logs/{UTC.localize(datetime.utcnow()).strftime("%Y%m%d_%H%M%S_%f")}'
-    log_format = "%(asctime)s_%(msecs)03d:%(levelname)s:%(funcName)s:%(message)s"
-
-    # Setup the logging module.
-    logging.basicConfig(
-        filename=filename,
-        format=log_format,
-        datefmt="%Y%m%d_%H%M%S",
-        level=logging.DEBUG,
-    )
-    logging.captureWarnings(True)
-    logging.Formatter.converter = time.gmtime
-
-    # Configure and add a console stream handler.
-    console_handler = logging.StreamHandler()
-    console_handler.setLevel(logging.INFO)
-    log_formatter = logging.Formatter(log_format)
-    console_handler.setFormatter(log_formatter)
-    logging.getLogger().addHandler(console_handler)
-
-    return filename
-
-
-def _setup_logger(name, propagate=True) -> logging.Logger:
-    """Sets up a logger object.
-
-    This function is intended to be used at the top-level of a module.
-
-    Parameters
-    ----------
-    name : str
-        Name of the file where this function is called.
-    propagate : bool, optional
-        Propogate this logger module's messages to the root logger or not, by
-        default True.
-
-    Returns
-    -------
-    logging.Logger
-        The logger.
-    """
+from datetime import datetime, timezone
+
+'''
+    ACCEPTS:
+        name=<anyname>
+        logfilename=<anypath/file>      [default = e2c_logs/dflt_log-{datetime.now(timezone.utc).strftime('%Y%m%d_%H%M%S_%f')}.log]
+        log_level=<level>               [default = DEBUG]
+        to_console=[True|False]         [default = False]
+        to_logfile=[True|False]         [default = False]
+        propagate=[True|False]          [default = False]
+'''
+
+default_log_dir = "e2c_logs"
+default_log = f"{default_log_dir}/e2c_root_log-{datetime.now(timezone.utc).strftime('%Y%m%d_%H%M%S_%f')}.log"
+
+def e2c_logger(name=None, logfilename=default_log, set_log_level=None, to_console=False, to_logfile=False, propagate=False):
+
+    # print(f"DEBUG: _logger: entered e2c_logger at {datetime.now(timezone.utc).strftime('%Y%m%d_%H%M%S_%f')} called by {name}", flush=True)
+
+    default_log_lvl = logging.DEBUG
+
+    # create logging directory as required
+
+    if to_logfile:
+        dn = os.path.dirname(logfilename)
+        if len(dn) and not os.path.exists(dn):
+            os.makedirs(dn)
+
     logger = logging.getLogger(name)
+    if name == None or name == "__main__":
+        logger = logger.root
+
     logger.propagate = propagate
 
+    if set_log_level == "None" or set_log_level == "DEBUG":
+        log_level = default_log_lvl
+    elif set_log_level == "INFO":
+        log_level = logging.INFO
+    elif set_log_level == "WARNING":
+        log_level = logging.WARNING
+    elif set_log_level == "ERROR":
+        log_level = logging.ERROR
+    elif set_log_level == "CRITICAL":
+        log_level = logging.CRITICAL
+    else: log_level = default_log_lvl
```

Comment:
For modules that do not "import logging", is "logging.<LEVEL_NAME>" available?  Or does it become available by importing our local "_logger", since THAT imports logging?

Reply:
You need to add the import at the top of each file that uses the logger. If that is more work than passing a string (e.g., "INFO"), then I'm fine with keeping that arg as a string.

Reply:
It is not that much work.  I don't know what "overhead" is involved with runtime (compile-time?) import of "logging", just to access a defined constant.  Six of One.  It would make the _logger cleaner.